src/plugins/workspace/public/components/data_source_association/association_data_source_modal.tsx

```diff
@@ -231,26 +254,68 @@ export const AssociationDataSourceModalContent = ({
   useEffect(() => {
     setIsLoading(true);
     getDataSourcesList(savedObjects.client, ['*'])
-      .then((dataSourcesList) => fetchDataSourceConnections(dataSourcesList, http, notifications))
-      .then((connections) => {
-        setAllConnections(connections);
+      .then((dataSourcesList) => {
+        const {
+          openSearchConnections,
+          dataConnections,
+        } = convertDataSourcesToOpenSearchAndDataConnections(dataSourcesList);
+        //  Only data source saved object type needs to fetch data source connections, data connection type object not, use loadingStatus to track the loading status of each data source connections
+        const initialLoadingStatus = [
+          ...openSearchConnections.map((ds) => ({ id: ds.id, status: true })),
+          ...dataConnections.map((ds) => ({ id: ds.id, status: false })),
+        ].reduce((acc, { id, status }) => {
+          acc[id] = status;
+          return acc;
+        }, {} as Record<string, boolean>);
+
+        setLoadingStatus(initialLoadingStatus);
+
+        // display data sources connections first while loading direct query connection
+        setAllConnections([...openSearchConnections, ...dataConnections]);
+        return { openSearchConnections };
+      })
+      .then(({ openSearchConnections }) => {
+        openSearchConnections.forEach((ds) => {
+          //  only fetch direct query connections for data source saved object, and set loading status accordingly
+          fetchDirectQueryConnectionsByIDs([ds.id], http, notifications)
+            .then((directQueryConnections) => {
+              setAllConnections((prev) => {
+                return prev.map((connection) => {
+                  if (connection.id === ds.id) {
+                    return {
+                      ...connection,
+                      relatedConnections: directQueryConnections,
+                    };
+                  }
+                  return connection;
+                });
+              });
+            })
+            .finally(() => {
+              setLoadingStatus((prev) => ({ ...prev, [ds.id]: false }));
+            });
+        });
       })
       .finally(() => {
         setIsLoading(false);
```

Comment:
In OpenSearch Data source case, the loading state will be set to false once the saved data sources being retrieved.
In Direct query connections case, should the loading state be set to false only after all the direct query connections being fetched?